### PR TITLE
feat: Add new pool initialization action and validation at PM call

### PR DESCRIPTION
**Pull request by ChefMist**
This PR does the following:

- adds pool initialization action (V4_CL_INITIALIZE_POOL and V4_BIN_INITIALIZE_POOL)
- for position manager call, ensure user's input is not using the blacklisted command which can result in potential loss of user funds

### CODE REVIEW

**Comment by ChefMist** [comment]
CI fail as expected due to exceed bytecode size. see https://github.com/pancakeswap/pancake-v4-universal-router/pull/27 on the passing PR

File: src/modules/V3ToV4Migrator.sol [hunk 1]
```diff
@@ -21,4 +38,77 @@ abstract contract V3ToV4Migrator is RouterImmutables {
...
+        // the second bytes(0) extracts the actions parameter from unlockData
+        bytes calldata actions = slice.toBytes(0).toBytes(0);
+
+        uint256 numActions = actions.length;
+
+        for (uint256 actionIndex = 0; actionIndex < numActions; actionIndex++) {
+            uint256 action = uint8(actions[actionIndex]);
+
+            if (
+                action == Actions.CL_INCREASE_LIQUIDITY || action == Actions.CL_DECREASE_LIQUIDITY
```
Comment (chefburger) [inline src/modules/V3ToV4Migrator.sol:77]: isBlacklistedCLPositionManagerAction ?

File: src/modules/V3ToV4Migrator.sol [hunk 2]
```diff
@@ -21,4 +38,77 @@ abstract contract V3ToV4Migrator is RouterImmutables {
...
+        // the first bytes(0) extracts the unlockData parameter from modifyLiquidities
+        // unlockData = `abi.encode(bytes actions, bytes[] params)`
+        // the second bytes(0) extracts the actions parameter from unlockData
+        bytes calldata actions = slice.toBytes(0).toBytes(0);
+
+        uint256 numActions = actions.length;
+
+        for (uint256 actionIndex = 0; actionIndex < numActions; actionIndex++) {
+            uint256 action = uint8(actions[actionIndex]);
+            if (action == Actions.BIN_REMOVE_LIQUIDITY) {
```
Comment (chefburger) [inline src/modules/V3ToV4Migrator.sol:109]: isBlacklistedBinPositionManagerAction ?

File: src/modules/V3ToV4Migrator.sol [hunk 3]
```diff
@@ -4,10 +4,27 @@ pragma solidity ^0.8.0;
...
 abstract contract V3ToV4Migrator is RouterImmutables {
+    using CalldataDecoder for bytes;
+
+    error NotAuthorizedForToken(uint256 tokenId);
+    error InvalidAction(bytes4 action);
+    error OnlyMintAllowed();
+    error OnlyAddLiqudityAllowed();
+    error BlacklistedAction();
+
+    enum PoolType {
```
Comment (chefburger) [inline src/modules/V3ToV4Migrator.sol:23]: Curious is it being used somewhere ?

File: src/modules/V3ToV4Migrator.sol [hunk 4]
```diff
@@ -4,10 +4,27 @@ pragma solidity ^0.8.0;
...
 /// @title V3 to V4 Migrator
 /// @notice A contract that migrates liquidity from PancakeSwap V3 to V4
 abstract contract V3ToV4Migrator is RouterImmutables {
+    using CalldataDecoder for bytes;
+
+    error NotAuthorizedForToken(uint256 tokenId);
+    error InvalidAction(bytes4 action);
+    error OnlyMintAllowed();
+    error OnlyAddLiqudityAllowed();
+    error BlacklistedAction();
```
Comment (chefburger) [inline src/modules/V3ToV4Migrator.sol:21]: This is duplicate with OnlyMintAllowed/OnlyAddLiqudityAllowed?

File: src/modules/V3ToV4Migrator.sol (hunk 3 above)
Comment (ChefMist) [reply src/modules/V3ToV4Migrator.sol:23]: my bad.. this was my POC code previously to see if it reduces the bytecode size.. will remove relevant and udpate shortly..

Comment (ChefMist) [reply src/modules/V3ToV4Migrator.sol]: updated, all the other comments should be 'resolved' now, sorry about it

**Review by chefburger — APPROVED** [review]
lgtm

File: src/modules/V3ToV4Migrator.sol [hunk 5]
```diff
@@ -21,4 +32,68 @@ abstract contract V3ToV4Migrator is RouterImmutables {
...
+        bytes calldata slice = inputs[4:];
+        // the first bytes(0) extracts the unlockData parameter from modifyLiquidities
+        // unlockData = `abi.encode(bytes actions, bytes[] params)`
+        // the second bytes(0) extracts the actions parameter from unlockData
+        bytes calldata actions = slice.toBytes(0).toBytes(0);
+
+        uint256 numActions = actions.length;
+
+        for (uint256 actionIndex = 0; actionIndex < numActions; actionIndex++) {
+            uint256 action = uint8(actions[actionIndex]);
```
Comment (jpopxfile) [inline src/modules/V3ToV4Migrator.sol:59]: Why casting uint8 here for a uint256 variable?

**Comment by jpopxfile** [comment]
https://github.com/pancakeswap/pancake-v4-universal-router/blob/25b499eea123f1334867eac49e43e3ff4e73df3e/src/libraries/Commands.sol#L42

Update comment.

File: src/modules/V3ToV4Migrator.sol (hunk 5 above)
Comment (ChefMist) [reply src/modules/V3ToV4Migrator.sol:59]: `actions[actionIndex]` returns a bytes1 and we need to explictly convert it to a `uint` type

**Pull request merged**